weedcoco/index/indexing.py

```diff
@@ -0,0 +1,140 @@
+            and "temp" not in os.listdir()
+        ):
+            self.temp_index_dir = "temp"
+            os.mkdir(self.temp_index_dir)
+
+    def modify_coco(self):
+        def _flatten(src, dst, prefix):
+            for k, v in src.items():
+                dst[f"{prefix}__{k}"] = v
+
+        with open(self.weedcoco_path) as f:
+            coco = json.load(f)
+        try:
+            del coco["info"]
+            del coco["collection_memberships"]
+        except KeyError:
+            pass
+
+        id_lookup = {}
+        for key, objs in coco.items():
+            for obj in objs:
+                id_lookup[key, obj["id"]] = obj
+
+        variable_to_null_fields = ["camera_fov", "camera_lens_focallength"]
+
+        for agcontext in coco["agcontexts"]:
+            # massage for ES
+            for field in variable_to_null_fields:
+                if agcontext.get(field) == "variable":
+                    del agcontext[field]
+
+        for annotation in coco["annotations"]:
+            image = id_lookup["images", annotation["image_id"]]
+            image.setdefault("annotations", []).append(annotation)
+            annotation["category"] = id_lookup["categories", annotation["category_id"]]
+            # todo: add collection, license
+            _flatten(annotation["category"], annotation, "category")
+            # todo: add collection from collection_memberships
+            image["thumbnail"] = str(
+                self.thumbnail_dir / os.path.basename(image["file_name"])
+            )
+
+        for image in coco["images"]:
+            try:
+                image["resolution"] = image["width"] * image["height"]
+            except KeyError:
+                pass
+            image["agcontext"] = id_lookup["agcontexts", image["agcontext_id"]]
+            image["sortKey"] = hash(
+                image["file_name"]
+            )  # for deterministic random order
+            _flatten(image["agcontext"], image, "agcontext")
+            # todo: add license
+            image["task_type"] = set()
+            for annotation in image["annotations"]:
+                for k in annotation:
+                    image.setdefault(f"annotation__{k}", []).append(annotation[k])
+
+                # determine available task types
+                image["task_type"].add("classification")
+                if "segmentation" in annotation:
+                    image["task_type"].add("segmentation")
+                    image["task_type"].add("bounding box")
+                if "bbox" in annotation:
+                    image["task_type"].add("bounding box")
+            image["task_type"] = sorted(image["task_type"])
+
+        self.coco = coco
+
+    def generate_batches(self):
+
+        coco = self.coco
+        if "images" in coco and len(coco["images"]) > 0:
+            for i in range(ceil(len(coco["images"]) / self.index_batch)):
+                with open(f"{self.temp_index_dir}/index{str(i+1)}.json", "w") as f:
```

Comment:
Are we writing to disk for debugging purposes, or some other reason? It seems strange to write to disk if we're only going to read it back in again right away, and into memory (rather than streaming from disk).

Reply:
We could just generate a string for each batch:
```py
for batch in image_batches:
    blobs = []
    for image in batch:
        blobs.append('{"index": {"_index": "weedid", "_type": "image"}}')
        blobs.append(json.dumps(out))
    yield "\n".join(blobs)
```
In this case the post method would take an iterable of strings to post.

I note that the [Python `elasticsearch` package](https://pypi.org/project/elasticsearch/) manages the batching of objects for bulk upload, but also uses multithreading to post them more efficiently.